src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4264,224 +4150,190 @@ .maxstack  3
+    IL_00bd:  ldarg.0
+    IL_00be:  ldfld      ""bool C.<Main>d__0.<>s__4""
+    IL_00c3:  brtrue     IL_0038
+    IL_00c8:  ldarg.0
+    IL_00c9:  ldnull
+    IL_00ca:  stfld      ""IMyAsyncEnumerator<int> C.<Main>d__0.<>s__2""
+    IL_00cf:  leave.s    IL_00eb
   }
   catch System.Exception
   {
     // sequence point: <hidden>
-    IL_01ac:  stloc.s    V_8
-    IL_01ae:  ldarg.0
-    IL_01af:  ldc.i4.s   -2
-    IL_01b1:  stfld      ""int C.<Main>d__0.<>1__state""
-    IL_01b6:  ldarg.0
-    IL_01b7:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder C.<Main>d__0.<>t__builder""
-    IL_01bc:  ldloc.s    V_8
-    IL_01be:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetException(System.Exception)""
-    IL_01c3:  nop
-    IL_01c4:  leave.s    IL_01da
+    IL_00d1:  stloc.s    V_4
+    IL_00d3:  ldarg.0
+    IL_00d4:  ldc.i4.s   -2
+    IL_00d6:  stfld      ""int C.<Main>d__0.<>1__state""
+    IL_00db:  ldarg.0
+    IL_00dc:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder C.<Main>d__0.<>t__builder""
+    IL_00e1:  ldloc.s    V_4
+    IL_00e3:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetException(System.Exception)""
+    IL_00e8:  nop
+    IL_00e9:  leave.s    IL_00ff
   }
   // sequence point: }
-  IL_01c6:  ldarg.0
-  IL_01c7:  ldc.i4.s   -2
-  IL_01c9:  stfld      ""int C.<Main>d__0.<>1__state""
+  IL_00eb:  ldarg.0
+  IL_00ec:  ldc.i4.s   -2
+  IL_00ee:  stfld      ""int C.<Main>d__0.<>1__state""
   // sequence point: <hidden>
-  IL_01ce:  ldarg.0
-  IL_01cf:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder C.<Main>d__0.<>t__builder""
-  IL_01d4:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()""
-  IL_01d9:  nop
-  IL_01da:  ret
+  IL_00f3:  ldarg.0
+  IL_00f4:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder C.<Main>d__0.<>t__builder""
+  IL_00f9:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()""
+  IL_00fe:  nop
+  IL_00ff:  ret
 }", sequencePoints: "C+<Main>d__0.MoveNext", source: source);
         }
 
+        [ConditionalFact(typeof(WindowsDesktopOnly), Reason = ConditionalSkipReason.NativePdbRequiresDesktop)]
+        public void TestWithInterfaceImplementingPattern_ChildImplementsDisposeAsync()
+        {
+            string source = @"
+using static System.Console;
+using System.Threading.Tasks;
+
+public interface ICollection<T>
+{
+    IMyAsyncEnumerator<T> GetAsyncEnumerator(System.Threading.CancellationToken token = default);
+}
+public interface IMyAsyncEnumerator<T>
+{
+    T Current { get; }
+    Task<bool> MoveNextAsync();
+}
+
+public class Collection<T> : ICollection<T>
+{
+    public IMyAsyncEnumerator<T> GetAsyncEnumerator(System.Threading.CancellationToken token = default)
+    {
+        return new MyAsyncEnumerator<T>();
+    }
+    public System.Threading.Tasks.ValueTask DisposeAsync()
```

Comment:
>DisposeAsync [](start = 44, length = 12)

Should this `DisposeAsync` be defined on `MyAsyncEnumerator<T>` rather than here?

Reply:
Good catch. Thanks